### ISSUE TITLE
[CHORE]: Openssl version prevents the creation of workspaces

### ISSUE THREAD

**Issue by ennob** [Mintplex-Labs/anything-llm#898]
### How are you running AnythingLLM?

AnythingLLM desktop app

### What happened?

I am using the AppImage for Linux. When attempting to create a new workspace I get the following error:

```
Error: Invalid `prisma.workspaces.create()` invocation: Prisma Client could not locate the Query Engine for runtime "debian-openssl-1.1.x".

This happened because Prisma Client was generated for "debian-openssl-3.0.x", but the actual deployment required "debian-openssl-1.1.x".

Add "debian-openssl-1.1.x" to `binaryTargets` in the "schema.prisma" file and run `prisma generate` after saving it: generator client { provider = "prisma-client-js" binaryTargets = ["native", "debian-openssl-1.1.x"] }

The following locations have been searched: /tmp/.mount_AnythiEIHOMn/resources/backend/node_modules/.prisma/client /tmp/.mount_AnythiEIHOMn/resources/backend/node_modules/@prisma/client /home/tim/Documents/anything-llm-desktop/anything-llm/server/node_modules/@prisma/client /tmp/prisma-engines /tmp/.mount_AnythiEIHOMn/resources/backend/prisma

```
It seems to be a problem with the openssl version, but it must be the version included in the appimage because my system is running openssl 3.0.10 (on Ubuntu)

```
> openssl version
OpenSSL 3.0.10 1 Aug 2023 (Library: OpenSSL 3.0.10 1 Aug 2023)
```

Maybe both versions could be supported for maximum flexibility?

### Are there known steps to reproduce?

1. Run appimage desktop application
2. Select ollama running on localhost
3. Use default embedding and database
4. Create workspace <-- this fails

To get a more verbose error message 

5. Restart the application which sends you directly to the main interface (on-boarding has already been completed). 
6. Create workspace <- This will fail again but with a more verbose error message.

**Comment by timothycarambat** [comment]
This is because we compile the AppImage on Ubuntu 22 and you must be on another version. This was resolved in a discord thead via manual recomp :/. Gonna leave this open as basically, we need to see if Prisma can generate  the client at runtime so the correct openssl is used.

https://discord.com/channels/1114740394715004990/1194390131247943690/1215776724243976222

Solution: https://discord.com/channels/1114740394715004990/1194390131247943690/1215964957099819049

**Comment by timothycarambat** [comment]
TLDR:
- cd to where ur AnythingLLMDesktop.AppImage is
- run ./AnythingLLMDesktop.AppImage --appimage-extract
- cd squashfs-root/resources/backend/
- `node node_modules/prisma generate` (will need node installed)
- cd ../../
- ./AppRun

_this is temporary_

**Comment by ennob** [comment]
Thanks, that works for me. I rebuilt the appimage with 
`env ARCH=x86_64 ./appimagetool-x86_64.AppImage ./squashfs-root AnythingLLMDesktop-modified.AppImage`
just for convenience.

**Comment by timothycarambat** [comment]
@ennob  Are you on an ARM machine?

**Comment by ennob** [comment]
No, x86_64 GNU/Linux (Ubuntu 23.10)

**Comment by timothycarambat** [comment]
Sorry i just realized that was a dumb comment 😆

**Comment by AndKe** [comment]
Exactly same issue here, (Also Ubuntu 23.10) - what are the chances of a fix & new release soon?
Thank you

**Comment by timothycarambat** [comment]
I am unable to replicate this issue on a totally fresh install of Ubuntu 22.0 LTS that the appimage was not built on.

If you are running into this issue - can you attempt to run this version (1.4.1) that basically pins the ENVs `PRISMA_SCHEMA_ENGINE_BINARY` & `PRISMA_QUERY_ENGINE_LIBRARY` to the local binaries bundled in the app instead of assuming the system.

https://s3.us-west-1.amazonaws.com/public.useanything.com/support/linux-prisma-patch/AnythingLLMDesktop.AppImage

**Comment by ennob** [comment]
@timothycarambat v1.4.1 works for me without any modification. v1.4.0 (from https://useanything.com/download) does not.

**Comment by timothycarambat** [comment]
@ennob Thank you for trying that out! Will wait for a couple more confirmations before merging but so far looks like. What OS are you on?

**Comment by Nindaleth** [comment]
I can confirm that the v1.4.1 prerelease linked above works out of the box. I'm on Fedora 39.

**Comment by timothycarambat** [comment]
Alright, gonna ship this then in the next patch. If this issue closes, the patch is released.

(if you are coming to this thread and it is open - please still test and confirm)

**Comment by AndKe** [comment]
@timothycarambat 1.4.1 works for mee too (ubuntu 23.10)  - thank you.

**Comment by timothycarambat** [comment]
New release is available on useanything.com/download and includes this patch + new custom folder management. You can download and use that image now with no issue

**Comment by fengyunzaidushi** [comment]
i have download the latest version,why i still have the problem:
![image](https://github.com/Mintplex-Labs/anything-llm/assets/53250064/b4380086-c1b5-4507-a0c1-a353ea6dad93)

**Comment by timothycarambat** [comment]
@Godot-Fye Since you have this issue and I don't have access to an environment to replicate I created a rebuild of v1.5.0 but with some changes to the build process for the AppImage to include the required prisma engine since the `/tmp` directory AppImages mount to is not writable (so the engine cannot download on the fly).

Can you test this for me? If it works for you it will work for everyone. Can confirm it works on Ubuntu 22LTS
https://s3.us-west-1.amazonaws.com/public.useanything.com/staging/AnythingLLMDesktop.AppImage

**Comment by Godot-Fye** [comment]
> @fengyunzaidushi | @Godot-Fye Since you have this issue and I don't have access to an environment to replicate I created a rebuild of v1.5.0 but with some changes to the build process for the AppImage to include the required prisma engine since the `/tmp` directory AppImages mount to is not writable (so the engine cannot download on the fly).
> 
> Can you test this for me? If it works for you it will work for everyone. Can confirm it works on Ubuntu 22LTS 

I have run the Version, but it "stops" on this screen (Version 1.4.4 loads fine): 

https://s3.us-west-1.amazonaws.com/public.useanything.com/staging/AnythingLLMDesktop.AppImage

It seems that it can't load my existent Worskpace. No Error Messages appear. Any log i can post to help you?

**Comment by timothycarambat** [comment]
When it mounts there is no logs at all from the app image?

**Comment by Godot-Fye** [comment]
> When it mounts there is no logs at all from the app image?

Where can i found the logs? :)

**Comment by timothycarambat** [comment]
When you run `./AnythingLLMDesktop.AppImage` the terminal window you are booting from will dump logs into that pane. How are you running the AppImage?

**Comment by zaro** [comment]
Happens to me also with Fedora 40 and 1.5.0 :

```
Error: Invalid `prisma.workspaces.create()` invocation: Prisma Client could not locate the Query Engine for runtime "rhel-openssl-1.1.x". This happened because Prisma Client was generated for "debian-openssl-3.0.x", but the actual deployment required "rhel-openssl-1.1.x". Add "rhel-openssl-1.1.x" to `binaryTargets` in the "schema.prisma" file and run `prisma generate` after saving it: generator client { provider = "prisma-client-js" binaryTargets = ["native", "rhel-openssl-1.1.x"] } The following locations have been searched: /tmp/.mount_AnythiAhwFTZ/resources/backend/node_modules/.prisma/client /tmp/.mount_AnythiAhwFTZ/resources/backend/node_modules/@prisma/client /home/tim/Documents/anything-llm-desktop/anything-llm/server/node_modules/@prisma/client /tmp/prisma-engines /tmp/.mount_AnythiAhwFTZ/resources/backend/prisma
```

**Comment by timothycarambat** [comment]
That new staging AppImage only includes the `openssl-` engines. Not `rhel` however seems like ill have to add all of them now

**Comment by Godot-Fye** [comment]
> When you run `./AnythingLLMDesktop.AppImage` the terminal window you are booting from will dump logs into that pane. How are you running the AppImage?

I just start the Image over the KDE Menu, it asks me if i wish to run it one time or integrate it in my system. If is start the Appimage on Console, this is the Log Output:

`Updating location Prisma binaries for linux builds.
libva error: vaGetDriverNames() failed with unknown libva error
Prisma schema loaded from prisma/schema.prisma
Datasource "db": SQLite database "anythingllm.db" at "file:/home/fye/.config/anythingllm-desktop/storage/anythingllm.db"

Already in sync, no schema change or pending migration was found.

EROFS: read-only file system, unlink '/tmp/.mount_AnythiNlbwGX/resources/backend/node_modules/.prisma/client/index.js'


Prisma schema loaded from prisma/schema.prisma
Error: 
EROFS: read-only file system, unlink '/tmp/.mount_AnythiNlbwGX/resources/backend/node_modules/.prisma/client/index.js'


[OllamaProcessManager] Ollama will bind on port 11434 when booted.
`

**Comment by timothycarambat** [comment]
@Godot-Fye but the app never loads? Looks like the DB loaded and was migrated.

**Comment by Godot-Fye** [comment]
> @Godot-Fye but the app never loads? Looks like the DB loaded and was migrated.

1.4.4 loads perfect, 1.5.0 does not. 1.5.0 "hangs" on the screen i postet :)

**Comment by fengyunzaidushi** [comment]
@timothycarambat 
i have download the v1.5.1
it still the same problem:

![image](https://github.com/Mintplex-Labs/anything-llm/assets/53250064/7517fb22-ef5c-4022-9b69-8bad9744b3cb)

```
The following locations have been searched:
  /tmp/.mount_Anythiz5DB7b/resources/backend/node_modules/.prisma/client
  /tmp/.mount_Anythiz5DB7b/resources/backend/node_modules/@prisma/client
  /home/tim/Documents/anything-llm-desktop/anything-llm/server/node_modules/@prisma/client
  /tmp/prisma-engines
  /tmp/.mount_Anythiz5DB7b/resources/backend/prisma
prisma:error 
Invalid `prisma.system_settings.findFirst()` invocation:


Prisma Client could not locate the Query Engine for runtime "debian-openssl-1.1.x".

This happened because Prisma Client was generated for "debian-openssl-3.0.x", but the actual deployment required "debian-openssl-1.1.x".
Add "debian-openssl-1.1.x" to `binaryTargets` in the "schema.prisma" file and run `prisma generate` after saving it:

generator client {
  provider      = "prisma-client-js"
  binaryTargets = ["native", "debian-openssl-1.1.x"]
}

The following locations have been searched:
  /tmp/.mount_Anythiz5DB7b/resources/backend/node_modules/.prisma/client
  /tmp/.mount_Anythiz5DB7b/resources/backend/node_modules/@prisma/client
  /home/tim/Documents/anything-llm-desktop/anything-llm/server/node_modules/@prisma/client
  /tmp/prisma-engines
  /tmp/.mount_Anythiz5DB7b/resources/backend/prisma
prisma:error 
Invalid `prisma.system_settings.upsert()` invocation:


Prisma Client could not locate the Query Engine for runtime "debian-openssl-1.1.x".

This happened because Prisma Client was generated for "debian-openssl-3.0.x", but the actual deployment required "debian-openssl-1.1.x".
Add "debian-openssl-1.1.x" to `binaryTargets` in the "schema.prisma" file and run `prisma generate` after saving it:

generator client {
  provider      = "prisma-client-js"
  binaryTargets = ["native", "debian-openssl-1.1.x"]
}

The following locations have been searched:
  /tmp/.mount_Anythiz5DB7b/resources/backend/node_modules/.prisma/client
  /tmp/.mount_Anythiz5DB7b/resources/backend/node_modules/@prisma/client
  /home/tim/Documents/anything-llm-desktop/anything-llm/server/node_modules/@prisma/client
  /tmp/prisma-engines
  /tmp/.mount_Anythiz5DB7b/resources/backend/prisma
[TELEMETRY SENT] {
  event: 'workspace_created',
  distinctId: '85784926-7ce5-408a-bad9-4ce328fc82a2',
  properties: {
    multiUserMode: false,
    LLMSelection: 'localai',
    VectorDbSelection: 'lancedb',
    runtime: 'desktop'
  }
}
prisma:error 
Invalid `prisma.event_logs.create()` invocation:


Prisma Client could not locate the Query Engine for runtime "debian-openssl-1.1.x".

This happened because Prisma Client was generated for "debian-openssl-3.0.x", but the actual deployment required "debian-openssl-1.1.x".
Add "debian-openssl-1.1.x" to `binaryTargets` in the "schema.prisma" file and run `prisma generate` after saving it:

generator client {
  provider      = "prisma-client-js"
  binaryTargets = ["native", "debian-openssl-1.1.x"]
}

The following locations have been searched:
  /tmp/.mount_Anythiz5DB7b/resources/backend/node_modules/.prisma/client
  /tmp/.mount_Anythiz5DB7b/resources/backend/node_modules/@prisma/client
  /home/tim/Documents/anything-llm-desktop/anything-llm/server/node_modules/@prisma/client
  /tmp/prisma-engines
  /tmp/.mount_Anythiz5DB7b/resources/backend/prisma
```
thank you !

**Comment by nmcbride** [comment]
Same in Fedora.

@timothycarambat I've seen in a few places that you only have ubuntu and do not have access to other distros to test...
Virtual machines and containers are free. You can easily spin up these other environments to test. You can also automate it here on Github or any other similar CI tool.

**Comment by imbev** [comment]
Also encountering this problem:

```
Error: Invalid `prisma.workspaces.create()` invocation: Prisma Client could not locate the Query Engine for runtime "rhel-openssl-3.0.x". This happened because Prisma Client was generated for "debian-openssl-3.0.x", but the actual deployment required "rhel-openssl-3.0.x". Add "rhel-openssl-3.0.x" to `binaryTargets` in the "schema.prisma" file and run `prisma generate` after saving it: generator client { provider = "prisma-client-js" binaryTargets = ["native", "rhel-openssl-3.0.x"] } The following locations have been searched: /tmp/.mount_gearleAdjXS9/resources/backend/node_modules/.prisma/client /tmp/.mount_gearleAdjXS9/resources/backend/node_modules/@prisma/client /home/tim/Documents/anything-llm-desktop/anything-llm/server/node_modules/@prisma/client /tmp/prisma-engines /tmp/.mount_gearleAdjXS9/resources/backend/prisma
```

**Comment by craighay** [comment]
the temporary extract-and-rebuild workaround errors with 'Unexpected token':  (SUSE Linux Enterprise Server 12 SP5)

squashfs-root/resources/backend/node_modules/prisma/build/index.js:867
    if (typeof process !== "undefined" && typeof process.stderr?.write !== "function") {
                                                                ^
SyntaxError: Unexpected token .
    at createScript (vm.js:80:10)
    at Object.runInThisContext (vm.js:139:10)
    at Module._compile (module.js:607:28)
    at Object.Module._extensions..js (module.js:654:10)
    at Module.load (module.js:556:32)
    at tryModuleLoad (module.js:499:12)
    at Function.Module._load (module.js:491:3)
    at Function.Module.runMain (module.js:684:10)
    at startup (bootstrap_node.js:187:16)
    at bootstrap_node.js:608:3

**Comment by nmcbride** [comment]
That's not a good workaround anyway imo.
If I wanted to go through all that I'd just run from source. 

We just need a fixed build.

**Comment by studiosusmed** [comment]
> > @Godot-Fye but the app never loads? Looks like the DB loaded and was migrated.
> 
> 1.4.4 loads perfect, 1.5.0 does not. 1.5.0 "hangs" on the screen i postet :)

I deleted the config from /home/username/.config/anythingllm, after that it worked normally.

**Comment by nmcbride** [comment]
> I deleted the config from /home/username/.config/anythingllm, after that it worked normally.

I'm glad it is working for you. You must be on a debian-based or debian-like distro. 

For lots of other distros though like rhel-based (fedora) it will not work at all. I am on 1.5.4 and it gives the errors I show above because the appimage is only built with certain debian-based dependencies.

> Error: Invalid `prisma.workspaces.create()` invocation: Prisma Client could not locate the Query Engine for runtime 
"rhel-openssl-3.0.x". This happened because Prisma Client was generated for "debian-openssl-3.0.x", but the actual 
deployment required "rhel-openssl-3.0.x". Add "rhel-openssl-3.0.x" to `binaryTargets` in the "schema.prisma" file and 
run `prisma generate` after saving it:

**Comment by fedevx** [comment]
This bug affects Arch as well. Had to do the workaround mentioned in https://github.com/Mintplex-Labs/anything-llm/issues/898#issuecomment-1994867745 and https://github.com/Mintplex-Labs/anything-llm/issues/898#issuecomment-1995025128

**Comment by airtonix** [comment]
![image](https://github.com/Mintplex-Labs/anything-llm/assets/61225/49e90ea6-e264-4a16-a629-9a5eafb0f710)

**Comment by wencan** [comment]
AnythingLLM v1.5.5
AppImage

OS: Fedora Linux 40 (Workstation Edition) x86_64 
Host: 21D0 ThinkBook 14 G4+ ARA 
Kernel: 6.8.9-300.fc40.x86_64 
Uptime: 1 hour, 50 mins 
Packages: 2806 (rpm), 38 (flatpak) 
Shell: bash 5.2.26 
Resolution: 2880x1800 
DE: GNOME 46.1 
WM: Mutter 
WM Theme: Adwaita 
Theme: Adwaita [GTK2/3] 
Icons: Adwaita [GTK2/3] 
Terminal: gnome-terminal 
CPU: AMD Ryzen 5 6600H with Radeon Graphics (12) @ 4.564GHz 
GPU: AMD ATI Radeon 680M 
Memory: 5501MiB / 13649MiB 


![Screenshot from 2024-05-24 16-11-58](https://github.com/Mintplex-Labs/anything-llm/assets/1072917/a4481461-cbe2-4409-bb99-00883210ed14)
```
Error: Invalid `prisma.workspaces.create()` invocation: Prisma Client could not locate the Query Engine for runtime "rhel-openssl-1.1.x". This happened because Prisma Client was generated for "debian-openssl-3.0.x", but the actual deployment required "rhel-openssl-1.1.x". Add "rhel-openssl-1.1.x" to `binaryTargets` in the "schema.prisma" file and run `prisma generate` after saving it: generator client { provider = "prisma-client-js" binaryTargets = ["native", "rhel-openssl-1.1.x"] } The following locations have been searched: /tmp/.mount_AnythilkhXhf/resources/backend/node_modules/.prisma/client /tmp/.mount_AnythilkhXhf/resources/backend/node_modules/@prisma/client /home/tim/Documents/anything-llm-desktop/anything-llm/server/node_modules/@prisma/client /tmp/prisma-engines /tmp/.mount_AnythilkhXhf/resources/backend/prisma


```

**Comment by Milor123** [comment]
> TLDR:
> 
>     * cd to where ur AnythingLLMDesktop.AppImage is
> 
>     * run ./AnythingLLMDesktop.AppImage --appimage-extract
> 
>     * cd squashfs-root/resources/backend/
> 
>     * `node node_modules/prisma generate` (will need node installed)
> 
>     * cd ../../
> 
>     * ./AppRun
> 
> 
> _this is temporary_

This worked for me today, it was very rare because it had never happened to me, it always worked normally and suddenly opened it and stopped working.

But why does this happen?A few months have passed and what the error in the last version seems to be followed, I am using Manjaro.

**Comment by richardstevenhack** [comment]
Same problem on latest openSUSE Tumbleweed with latest AnyLLM AppImage from Web site.
I will try the workaround and report back.

**Comment by drAndric** [comment]
it continues to be an issue, v. 1.5.6 on Fedora 40 

> Error: Invalid `prisma.workspaces.create()` invocation: Prisma Client could not locate the Query Engine for runtime "rhel-openssl-3.0.x". This happened because Prisma Client was generated for "debian-openssl-3.0.x", but the actual deployment required "rhel-openssl-3.0.x". Add "rhel-openssl-3.0.x" to `binaryTargets` in the "schema.prisma" file and run `prisma generate` after saving it: generator client { provider = "prisma-client-js" binaryTargets = ["native", "rhel-openssl-3.0.x"] } The following locations have been searched: /tmp/.mount_AnythisV6KTZ/resources/backend/node_modules/.prisma/client /tmp/.mount_AnythisV6KTZ/resources/backend/node_modules/@prisma/client /home/tim/Documents/anything-llm-desktop/anything-llm/server/node_modules/@prisma/client /tmp/prisma-engines /tmp/.mount_AnythisV6KTZ/resources/backend/prisma`

**Comment by turcol** [comment]
The issue continues on v1.5.7 on Ubuntu 22.04.3 LTS while creating a new Workspace

> Error: Invalid `prisma.workspaces.create()` invocation: Prisma Client could not locate the Query Engine for runtime "debian-openssl-1.1.x". This happened because Prisma Client was generated for "debian-openssl-3.0.x", but the actual deployment required "debian-openssl-1.1.x". Add "debian-openssl-1.1.x" to `binaryTargets` in the "schema.prisma" file and run `prisma generate` after saving it: generator client { provider = "prisma-client-js" binaryTargets = ["native", "debian-openssl-1.1.x"] } The following locations have been searched: /tmp/.mount_AnythiqBOiNq/resources/backend/node_modules/.prisma/client /tmp/.mount_AnythiqBOiNq/resources/backend/node_modules/@prisma/client /home/tim/Documents/anything-llm-desktop/anything-llm/server/node_modules/@prisma/client /tmp/prisma-engines /tmp/.mount_AnythiqBOiNq/resources/backend/prisma

**Comment by richardstevenhack** [comment]
Confirm still an issue on 1.5.7 on openSUSE Tumbleweed.

**Comment by richardstevenhack** [comment]
OK, today I uninstalled the app, deleted all the configs, reinstalled from latest AppImage.

Went through the initial setup - got "failed to create workspace - undefined".

I assume it's still the prism issue. Is this going to be fixed any time soon? If not, I'm going to leave AnythingLLM aside and just work with some of my LLM GUIs that do work.

**Comment by nmcbride** [comment]
> I assume it's still the prism issue. Is this going to be fixed any time soon? If not, I'm going to leave AnythingLLM aside and just work with some of my LLM GUIs that do work.

I ended up also leaving in favor of: https://msty.app/.

**Comment by timothycarambat** [comment]
Locking the thread for now, we **know** this is an issue for linux users. Currently, there are two paths around this:
- Unbundle and bundle  https://github.com/Mintplex-Labs/anything-llm/issues/898#issuecomment-2144068919
- Use Docker (you are on Linux anyway?)

The reason is again - mounting of an AppImage creates a tmp directory that is write-only and this blocks the Prisma engines from running to migrate the databases as well as create the client so the creation calls fail

**Comment by timothycarambat** [comment]
Solved: https://docs.useanything.com/installation/desktop/linux

If you have issues with this installer script or otherwise that is a new issue going forward
